src/chpasswd.8
.\" -*- nroff -*-
.\" Copyright (C) 2003, 2004 Thorsten Kukuk
.\" Author: Thorsten Kukuk <kukuk@suse.de>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2 as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.TH chpasswd 8 "Feburary 2004" "pwdutils"
.SH NAME
chpasswd \- change user passwords in batch
.SH SYNOPSIS
\fBchpasswd\fR [\fB-D \fIbinddn\fR] [\fB-P \fIpath\fR] [\fB-c \fIdes\fR|\fImd5\fR|\fIblowfish\fR | \fB-e\fR] \fI[file]\fR
.SH DESCRIPTION
\fBchpasswd\fR changes passwords for user accounts in batch. It
reads a list of login and password pairs from standard input
or a file and uses this information to update the passwords of this
user accounts. The named account must exist and the password age will
be updated. Each input line is of the format:
.sp 1
          \fIuser_name\fR:\fIpassword\fR
.sp 1
If the hash algorithmus is not given on the commandline, the value
of \fIGROUP_CRYPT\fR or, if not specified, \fICRYPT\fR from
\fI/etc/default/passwd\fR is used as hash algorithmus.
If not configured, the traditinal \fIdes\fR algorithmus is used.
.SH OPTIONS
.TP
.BI "\-c" " des|md5|blowfish"
This option specifies the hash algorithmus, which should be used to
encrypt the passwords.
.TP
.B "\-e"
The passwords are expected to be in encrypted form. Normally the
passwords are expected to be cleartext.
.TP
.BI "\-D, \-\-binddn" " binddn"
Use the Distinguished Name binddn to bind to the LDAP directory.
The user will be prompted for a password for simple authentication.
.TP
.BI "\-P, \-\-path" " path"
The \fIpasswd\fR and \fIshadow\fR files are located below
the specified directory path. \fBchpasswd\fR will use this files,
not \fI/etc/passwd\fR and \fI/etc/shadow\fR.
.SH FILES
/etc/default/passwd \- default values for password hash
.SH SEE ALSO
.BR passwd (1),
.BR passwd (5),
.BR shadow (5)
.SH AUTHOR
Thorsten Kukuk <kukuk@suse.de>


NEWS
pwdutils NEWS -- history of user-visible changes.

Copyright (C) 2002-2012 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@thkukuk.de>.

Version 3.2.19
* Fix passwd -S

Version 3.2.18
* Fix changing of permissions if uid/gid of user will be modified

Version 3.2.17
* Fix useradd.local

Version 3.2.16
* Make sha512 new default password hash

Version 3.2.15
* newgrp: honour dynamically assigned groups

Version 3.2.14
* pwck: check +/- entries, too
* grpck: check */- entries, too
* login.defs: remove LOGIN_RETRIES
 
Version 3.2.13
* Only copy regular files if usermod moves home directory

Version 3.2.12
* Make sure modified group files are written to disk
* Update translations for sv, fr

Version 3.2.11
* Update translations
* Add translation for nl

Version 3.2.10
* Fix encoding of polish translations
* Fix long options of grpck/pwck
 
Version 3.2.9
* Move UMASK from /etc/login.defs to /etc/default/useradd

Version 3.2.8
* Fix usermod manual page
* Add quiet option to pam_rpasswd.so

Version 3.2.7
* Fix possible endless loop in userdel

Version 3.2.6
* Fix rpasswd protocol error
* Update translations

Version 3.2.5
* Add translations for da
* pwconv: Remove temporary files in case of an error

Version 3.2.4
* Make sure changes are really written to disk

Version 3.2.3
* New translations
* Adjust documentation

Version 3.2.2
* Don't lock system accounts by default

Version 3.2.1
* Adjust for new libxcrypt

Version 3.2.0
* Better support for audit plugin
* --preferred-uid/gid option for useradd and groupadd

Version 3.1.4
* Update polish translation.
* Remove FTMP references.
* Better pwconv support for  NIS entries

Version 3.1.3
* Allow compiling without native language support.
* Update swedish translation.

Version 3.1.2
* Check if crontab is installed before calling it.
* Fix seg.fault in pwconv in error case.

Version 3.1.1
* Adjust PAM config files

Version 3.1.0
* gpasswd: Add -l and -u option to lock/unlock group passwords
* passwd: Add support for BioAPI
* Add LDAP support for RFC2307bis

Version 3.0.7
* passwd.1: correct manual page
* passwd: Allow changing passwords with PAM even if service is unknown

Version 3.0.6
* chfn: Fix syntax check for home phone number

Version 3.0.5
* Fix compilation error of PAM module with new Linux-PAM release
* Fix CHARACTER_CLASS in /etc/login.defs

Version 3.0.4
* Fix pam_rpasswd.so

Version 3.0.3
* usermod: Adjust sp_lstchg if -p option is used
* useradd: Preserve existing home directory
* chage: Fix off-by-one errors in output

Version 3.0.2
* Fix loading of logging plugins
* Unify options (use always --service)
* Add -a option to usermod for shadow compatibility

Version 3.0.1
* usermod: Don't remove old home directory if an error occured

Version 3.0
* Add --stdin option for passwd
* Fix useradd -D option
* Fix PAM/syslog interaction
* Fix lot of ignored failing syscalls
* Compile daemons with -fpie/-pie
* Add support for logging plugins
* Add plugin for LAuS support
* Add plugin for audit support
* Add plugin to log with syslog
* Use libnscd
* Add support for GnuTLS
* groupmod: Add options to remove/add single user from a group
* usermod: Add options to remove/add single user from a list of groups
* rpasswd/rpasswdd: Better SLP support

Version 2.6.4
* rpasswd/pam_rpasswd.so: Fix SLP URL parsing

Version 2.6.3
* newgrp: Add -c option to allow sg emulation

Version 2.6.2
* Fix compile problem if SELinux is disabled
* rpasswdd: Re-register SLP service after timeout

Version 2.6.1
* newgrp: Fix NGROUPS_MAX handling

Version 2.6
* Add -f compat flag to groupadd
* Add UTF8/local locale translation
* Make regex for allowed accounts/groups configurable
* Add optional pam_rpasswd module
* Add SLP support to rpasswd/rpasswdd
* Add pwck
* Add grpck
* Add newgrp
* Add /etc/default/passwd support

Version 2.5.1
* Install all manual pages

Version 2.5
* Adjust for libselinux v1.5
* Add pwconv/pwunconv
* Add grpunconv

Version 2.4
* Fix -G handling of usermod
* Fix LDAP support
* Add gpasswd, groupmod and usermod
* Copy always all extended attributes
* Add SELinux support for passwd, chsh and chfn
* Make sure that group, passwd and shadow are always created with
  correct security attributes
* groupadd -g handling fixed
* vipw/vigr added
* Lot of bug fixes
* Add manual pages
* Add groupadd and groupdel
* useradd and userdel can now create/delete LDAP accounts
* Make compile with OpenLDAP 2.0.x again
* Rewrite ldap functions to meet demands
* chpasswd can now change LDAP passwords
* useradd: Lot of new error checks and bug fixes
* chfn/chsh: Allow user to abort with Ctrl-C during data input
* Add test suite

Version 2.3
* Add chpasswd
* Add expiry
* Fix chage options
* rpasswd client now checks server certificate

Version 2.2
* Add "-D binddn" option, to allow modifying LDAP data with a special
  account.
* passwd and chage: Allow modifying of password expiry informations
  stored in a LDAP database.

Version 2.1
* Add --disable-ldap option to configure
* Fix typos in chage.1 manual page
* Rewrite passwd manual page and explain all options

Version 2.0
* Fix -d and -E option of chage
* Use port 774 as new default
* Make compile with glibc 2.3 and gcc 2.95/3.2/3.3

Version 1.99
* Add chage
* Allow modifying of passwd/shadow file not in /etc (only chage yet)

Version 1.98
* Add manual pages for rpasswd and rpasswdd

Version 1.97
* rpasswd now reads config file with server name and port number
* Security fix for rpasswdd: don't allow remote changing of root
  password without authentification
* Add manual page for rpasswd.conf

Version 1.96
* Make compile with gcc 2.x
* root can now change passwords for other user
* timeout if client does not answer fast enough
* Sent LANG as locale to server

Version 1.95
* Make location of ldap config files compile time configurable

Version 1.94
* Add more sanity checks to prevent file corruption

Version 1.93
* Minor bug fixes to rpasswdd

Version 1.92

src/chpasswd.c
/* Copyright (C) 2003, 2004, 2005 Thorsten Kukuk
   Author: Thorsten Kukuk <kukuk@suse.de>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <time.h>
#include <fcntl.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#if defined(HAVE_XCRYPT_H)
#include <xcrypt.h>
#elif defined(HAVE_CRYPT_H)
#include <crypt.h>
#endif
#include <sys/stat.h>
#include <sys/types.h>
#ifdef HAVE_LIBNSCD_H
#include <libnscd.h>
#endif

#include "i18n.h"
#include "public.h"
#include "logging.h"
#include "logindefs.h"
#include "read-files.h"
#include "error_codes.h"
#include "parse_crypt_arg.h"

#ifdef USE_LDAP
#include "libldap.h"
#endif /* USE_LDAP */

#ifndef _
#define _(String) gettext (String)
#endif

static void
print_usage (FILE *stream, const char *program)
{
  fprintf (stream, _("Usage: %s [-D binddn] [-P path] [-e] [-c des|md5|blowfish] [file]\n"),
           program);
}

static void
print_help (const char *program)
{
  print_usage (stdout, program);
  fprintf (stdout, _("%s - update password entries in batch\n\n"), program);

#ifdef USE_LDAP
  fputs (_("  -D binddn      Use dn \"binddn\" to bind to the LDAP directory\n"),
         stdout);
#endif
  fputs (_("  -P path        Search passwd and shadow file in \"path\"\n"),
         stdout);
  fputs (_("  -c, --crypt    Password should be encrypted with DES, MD5 or blowfish\n"),
	 stdout);
  fputs (_("  -e, --encrypted The passwords are in encrypted form\n"),
	 stdout);
  fputs (_("  --service srv  Use nameservice 'srv'\n"), stdout);
  fputs (_("      --help     Give this help list\n"), stdout);
  fputs (_("  -u, --usage    Give a short usage message\n"), stdout);
  fputs (_("  -v, --version  Print program version\n"), stdout);
  fputs (_("Valid services are: files, nis, nisplus, ldap\n"), stdout);
}

int
main (int argc, char *argv[])
{
  FILE *input = NULL;
  const char *program = "chpasswd";
  crypt_t use_crypt;
  int encrypted = 0;
  char *buf = NULL;
  size_t buflen = 0;
  unsigned long line = 0, errors = 0;
  char *use_service = NULL;
#ifdef USE_LDAP
  char *oldclearpwd = NULL;
  char *binddn = NULL;
#endif

#ifdef ENABLE_NLS
  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);
#endif

  open_sec_log (program);

  use_crypt = parse_crypt_arg (getlogindefs_str ("CRYPT", "des"));

  while (1)
    {
      int c;
      int option_index = 0;
      static struct option long_options[] = {
#ifdef USE_LDAP
        {"binddn",  required_argument, NULL, 'D' },
#endif
        {"path",    required_argument, NULL, 'P' },
        {"crypt",   no_argument, NULL, 'c' },
        {"md5",     no_argument, NULL, 'm' },
        {"encrypt", no_argument, NULL, 'e' },
        {"service", required_argument, NULL, '\254' },
        {"version", no_argument, NULL, 'v' },
        {"usage",   no_argument, NULL, 'u' },
        {"help",    no_argument, NULL, '\255' },
        {NULL,      0,           NULL, '\0'}
      };

      c = getopt_long (argc, argv, "D:P:c:mevu",
                       long_options, &option_index);
      if (c == (-1))
        break;
      switch (c)
	{
#ifdef USE_LDAP
	case 'D':
	  binddn = optarg;
	  break;
#endif
	case 'P':
	  files_etc_dir = strdup (optarg);
          break;
	case 'c':
	  use_crypt = parse_crypt_arg (optarg);
	  break;
	case 'm':
	  use_crypt = MD5;
	  break;
	case 'e':
	  ++encrypted;
	  break;
        case '\254':
          if (use_service != NULL)
            {
              print_usage (stderr, program);
              return E_BAD_ARG;
            }

          if (strcasecmp (optarg, "yp") == 0 ||
              strcasecmp (optarg, "nis") == 0)
            use_service = "nis";
          else if (strcasecmp (optarg, "nis+") == 0 ||
                   strcasecmp (optarg, "nisplus") == 0)
            use_service = "nisplus";
          else if (strcasecmp (optarg, "files") == 0)
            use_service = "files";
#ifdef USE_LDAP
          else if (strcasecmp (optarg, "ldap") == 0)
            use_service = "ldap";
#endif
          else
            {
              fprintf (stderr, _("Service `%s' not supported.\n"), optarg);
              print_usage (stderr, program);
              return E_BAD_ARG;
            }
	  break;
	case '\255':
          print_help (program);
          return 0;
        case 'v':
          print_version (program, "2005");
          return 0;
        case 'u':
          print_usage (stdout, program);
	  return 0;
	default:
	  print_error (program);
	  return E_USAGE;
	}
    }

  argc -= optind;
  argv += optind;

  if (argc == 0)
    input = stdin;
  else if (argc == 1)
    {
      input = fopen (argv[0], "r");
      if (input == NULL)
	{
	  fprintf (stderr, "%s: %s: %s\n", program, argv[0],
		   strerror (errno));
	  return E_BAD_ARG;
	}
    }
  else if (argc > 1)
    {
      fprintf (stderr, _("%s: Too many arguments.\n"), program);
      print_error (program);
      return E_USAGE;
    }

#ifdef USE_LDAP
  if (binddn)
    {
      /* A user tries to change data stored in a LDAP database and
	 knows the Manager dn, now we need the password from him.  */
      ldap_session_t *session = create_ldap_session (LDAP_PATH_CONF);
      char *cp;

      if (session == NULL)
	return E_FAILURE;

      cp = getpass (_("Enter LDAP Password:"));

      if (open_ldap_session (session) != 0)
	return E_FAILURE;

      if (ldap_authentication (session, NULL, binddn, cp) != 0)
	return E_NOPERM;

      close_ldap_session (session);

      oldclearpwd = strdup (cp);
    }
#endif /* USE_LDAP */

  /* Read each line, separating login from the password. The password
     entry for each user will be looked up in the appropriate place,
     defined through the search order in /etc/nsswitch.conf.  */

  while (!feof (input))
    {
      char *tmp, *cp;
      user_t *pw_data;
      time_t now;
#if defined(HAVE_GETLINE)
      ssize_t n = getline (&buf, &buflen, input);
#elif defined (HAVE_GETDELIM)
      ssize_t n = getdelim (&buf, &buflen, '\n', input);
#else
      ssize_t n;

      if (buf == NULL)
        {
          buflen = 8096;
          buf = malloc (buflen);
        }
      buf[0] = '\0';
      fgets (buf, buflen - 1, input);
      if (buf != NULL)
        n = strlen (buf);
      else
        n = 0;
#endif /* HAVE_GETLINE / HAVE_GETDELIM */

      ++line;
      cp = buf;

      if (n < 1)
	break;

      tmp = strchr (cp, ':');
      if (tmp)
	*tmp = '\0';
      else
	{
	  fprintf (stderr,_("%s: line %ld: missing new password\n"),
		   program, line);
	  ++errors;
	  continue;
	}

      pw_data = do_getpwnam (cp, use_service);
      if (pw_data == NULL || pw_data->service == S_NONE)
	{
	  fprintf (stderr, _("%s: line %ld: unknown user %s\n"),
		   program, line, cp);
	  ++errors;
	  continue;
	}

      cp = tmp+1;
      tmp = strchr (cp, '\n');
      if (tmp)
	*tmp = '\0';

      if (encrypted)
	pw_data->newpassword = strdup (cp);
      else
	{
          char *salt;
	  struct crypt_data output;
	  memset (&output, 0, sizeof (output));

	  switch (use_crypt)
	    {
	    case DES:
	      /* If we don't support passwords longer 8 characters,
		 truncate them */
	      if (strlen (cp) > 8)
		cp[8] = '\0';
	      salt =  make_crypt_salt ("", 0);
	      if (salt != NULL)
	        pw_data->newpassword = strdup (crypt_r (cp, salt, &output));
	      else
		{
		  fprintf (stderr, _("Cannot create salt for standard crypt"));
		  ++errors;
		  continue;
		}
	      free (salt);
	      break;

	    case MD5:
	      /* MD5 has a limit of 127 characters */
	      if (strlen (cp) > 127)
		cp[127] = '\0';
	      salt = make_crypt_salt ("$1$", 0);
	      if (salt != NULL)
		pw_data->newpassword = strdup (crypt_r (cp, salt, &output));
	      else
		{
		  fprintf (stderr, _("Cannot create salt for MD5 crypt"));
		  ++errors;
		  continue;
		}
	      free (salt);
	      break;
	    case BLOWFISH:
#if defined(HAVE_XCRYPT_GENSALT_R)
	      /* blowfish has a limit of 72 characters */
	      if (use_crypt == BLOWFISH && strlen (cp) > 72)
		cp[72] = '\0';
	      salt = make_crypt_salt ("$2a$", 0 /* XXX crypt_rounds */);
	      if (salt != NULL)
		pw_data->newpassword = strdup (crypt_r (cp, salt, &output));
	      else
		{
		  fprintf (stderr, _("Cannot create salt for blowfish crypt"));
		  ++errors;
		  continue;
		}
	      free (salt);
#endif
	      break;
	    default:
	      abort();
	    }
	}
      time (&now);
      pw_data->spn.sp_lstchg = (long int)now / (24L*3600L);
      pw_data->sp_changed = TRUE;

#ifdef USE_LDAP
      /* Add binddn data if user is stored in LDAP database and
	 we know the binddn.  */
      if (pw_data->service == S_LDAP)
	{
	  if (binddn)
	    pw_data->binddn = strdup (binddn);
	  if (oldclearpwd)
	    pw_data->oldclearpwd = strdup (oldclearpwd);
	}
#endif

      if (write_user_data (pw_data, 0) != 0)
	{
	  fprintf (stderr,
		   _("%s: line %ld: cannot update password entry\n"),
		   program, line);
	  ++errors;
	  continue;
	}

      free_user_t (pw_data);
    }

#ifdef HAVE_NSCD_FLUSH_CACHE
  nscd_flush_cache ("passwd");
#endif

  if (errors)
    {
      fprintf (stderr, _("%s: errors occurred, %ld passwords not updated\n"),
	       program, errors);
    }

  if (input != stdin)
    fclose (input);

  return E_SUCCESS;
}
